### PR TITLE
Vis point cloud

### PR DESCRIPTION
**Pull request by simbamford**
A point cloud visualiser for ntupleviz; Some other tidying up.

### CODE REVIEW

**Comment by arrenglover** [comment]
note @simbamford that this branch has conflicts that should be resolved. You should do a merge of master into your branch and then push again.

**Comment by simbamford** [comment]
Should be resolved now.

File: python/libraries/bimvee/geometry.py [hunk 1]
```diff
@@ -101,6 +101,18 @@ def pose6qInterp(poseDict, time):
...
+    v1 = q1[1:4]
+    w2 = q2[0]
+    v2 = q2[1:4]
+    wOut = w1*w2 - np.dot(v1.T, v2)
+    vOut = w1*v2 + w2*v1 + np.dot(v1, v2)
+    qOut = np.zeros_like(q1)
+    qOut[0] = wOut
+    qOut[1:4] = vOut
+    return qOut
+
```
Comment (ghoshsuman) [inline python/libraries/bimvee/geometry.py:115]: The output quaternion here is equivalent to the rotation by quaternion q1, followed by rotation by quaternion q2, right? If so, then the output seems incorrect. It should be equivalent to the quaternion product (q2 . q1). I called it with some different rotations (unit quaternions) and the superposition principle doesn't hold for the outputted quaternion.

File: python/libraries/bimvee/importRpgDvsRos.py [hunk 2]
```diff
@@ -89,11 +91,13 @@ def bimveeTypeForRosbagType(rosbagType):
     if rosbagType == 'esim_msgs_OpticFlow': return 'flowMap'
     if rosbagType == 'geometry_msgs_PoseStamped': return 'pose6q'
     if rosbagType == 'geometry_msgs_Transform': return 'pose6q'
+    if rosbagType == 'geometry_msgs_TransformStamped': return 'pose6q'
     if rosbagType == 'geometry_msgs_TwistStamped': return 'twist'
     if rosbagType == 'sensor_msgs_CameraInfo': return 'cam'
     if rosbagType == 'sensor_msgs_Image': return 'frame'
     if rosbagType == 'sensor_msgs_Imu': return 'imu'
     if rosbagType == 'sensor_msgs_PointCloud2': return 'point3'
+    if rosbagType == 'tf_tfMessage': return 'pose6q'
```
Comment (ghoshsuman) [inline python/libraries/bimvee/importRpgDvsRos.py:100]: Don't think it should be decoded as 'pose6q'. There is a layer in between. geometry_msgs/TransformStamped is inside tf/tfMessage. In the case of ESIM, /tf has 2 geometry messages (1 for body and 1 for cam0).

File: python/libraries/bimvee/visualiser.py [hunk 3]
```diff
@@ -355,4 +355,128 @@ def get_dims(self):
...
+            # Move z out by 0.5, so that the data is between 0.5 and 1.5 distant in z
+            x = (x - 0.5) / (z + 0.5) + 0.5
+            y = (y - 0.5) / (z + 0.5) + 0.5
+        x *= smallestRenderDim
+        y *= smallestRenderDim
+        x = x.astype(int)
+        y = y.astype(int)
+        return x, y
+
+    def project_point(self, point, image, **kwargs):
```
Comment (ghoshsuman) [inline python/libraries/bimvee/visualiser.py:416]:
```suggestion
    def project_pointcloud_on_image(self, point, image, **kwargs):
```
Can we refactor function name to make the functionality clearer?

File: python/libraries/bimvee/visualiser.py [hunk 4]
```diff
@@ -355,4 +355,128 @@ def get_dims(self):
...
+        internalData = {'ts': data['ts'],
+                        'point': pointScaled
+                        }
+        self.__data = internalData
+            
+    def project3dTo2d(self, x=0, y=0, z=0, smallestRenderDim=1, **kwargs):
+        if kwargs.get('perspective', True):
+            # Move z out by 0.5, so that the data is between 0.5 and 1.5 distant in z
+            x = (x - 0.5) / (z + 0.5) + 0.5
+            y = (y - 0.5) / (z + 0.5) + 0.5
```
Comment (ghoshsuman) [inline python/libraries/bimvee/visualiser.py:409]: Could you please elaborate a bit what is happening here?

File: python/libraries/bimvee/visualiser.py [hunk 5]
```diff
@@ -355,4 +355,128 @@ def get_dims(self):
...
+        cosA = np.cos(roll)
+        cosB = np.cos(yaw)
+        cosC = np.cos(pitch)
+        sinA = np.sin(roll)
+        sinB = np.sin(yaw)
+        sinC = np.sin(pitch)
+        rotMat = np.array([[cosA*cosB, cosA*sinB*sinC-sinA*cosC, cosA*sinB*cosC+sinA*sinC],
+                           [sinA*cosB, sinA*sinB*sinC+cosA*cosC, sinA*sinB*cosC-cosA*sinC],
+                           [-sinB, cosB*sinC, cosB*cosC]], 
+                            dtype=np.float64)
```
Comment (ghoshsuman) [inline python/libraries/bimvee/visualiser.py:465]: Didn't verify the formulae. Assuming they are correct.

File: python/libraries/bimvee/geometry.py (hunk 1 above)
Comment (simbamford) [reply python/libraries/bimvee/geometry.py]: Good catch - I've corrected that.

Comment (simbamford) [reply python/libraries/bimvee/geometry.py]: (dot product instead of cross product).

File: python/libraries/bimvee/importRpgDvsRos.py (hunk 2 above)
Comment (simbamford) [reply python/libraries/bimvee/importRpgDvsRos.py:100]: Thanks for this observation - I've fixed this in importRosbag repo - now frame_id and child_frame_id are also imported and they can be used to separate out the poses for different bodies in a post-processing step if desired, similarly to what happens in importVicon.

File: python/libraries/bimvee/visualiser.py (hunk 4 above)
Comment (simbamford) [reply python/libraries/bimvee/visualiser.py:409]: Added a comment - also totally refactored :)

File: python/libraries/bimvee/visualiser.py (hunk 3 above)
Comment (ghoshsuman) [reply python/libraries/bimvee/visualiser.py]: Refactored in new commit.

**Pull request merged**